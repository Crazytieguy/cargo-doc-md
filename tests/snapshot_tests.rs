use cargo_doc_md::{converter, parser, ConversionOptions};
use std::path::Path;

#[test]
fn test_multifile_output_structure() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let mut file_paths: Vec<_> = output.files.keys().collect();
    file_paths.sort();

    insta::assert_debug_snapshot!("multifile_structure", file_paths);
    assert_eq!(output.crate_name, "test_crate");
}

#[test]
fn test_index_file() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let index_content = output.files.get("index.md").expect("index.md not found");
    insta::assert_snapshot!("index_md", index_content);
}

#[test]
fn test_lib_module() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    if let Some(lib_content) = output.files.get("test_crate.md") {
        insta::assert_snapshot!("lib_module", lib_content);
    }
}

#[test]
fn test_types_module() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    if let Some(types_content) = output.files.get("types.md") {
        insta::assert_snapshot!("types_module", types_content);
    }
}

#[test]
fn test_functions_module() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    if let Some(functions_content) = output.files.get("functions.md") {
        insta::assert_snapshot!("functions_module", functions_content);
    }
}

#[test]
fn test_nested_modules() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    if let Some(nested_content) = output.files.get("nested.md") {
        insta::assert_snapshot!("nested_module", nested_content);
    }

    if let Some(inner_content) = output.files.get("nested/inner.md") {
        insta::assert_snapshot!("nested_inner_module", inner_content);
    }

    if let Some(deep_content) = output.files.get("nested/inner/deep.md") {
        insta::assert_snapshot!("nested_deep_module", deep_content);
    }
}

#[test]
fn test_struct_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("## test_crate::UnitStruct"));
    assert!(lib_content.contains("**Unit Struct**"));

    assert!(lib_content.contains("## test_crate::TupleStruct"));
    assert!(lib_content.contains("**Tuple Struct**"));

    assert!(lib_content.contains("## test_crate::PlainStruct"));
    assert!(lib_content.contains("**Fields:**"));
}

#[test]
fn test_generic_struct_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("## test_crate::GenericStruct"));
    assert!(lib_content.contains("**Generic Parameters:**"));
}

#[test]
fn test_enum_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("## test_crate::SimpleEnum"));
    assert!(lib_content.contains("**Variants:**"));

    assert!(lib_content.contains("## test_crate::ComplexEnum"));
}

#[test]
fn test_function_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let functions_content = output.files.get("functions.md").expect("functions module not found");

    assert!(functions_content.contains("fn add"));
    assert!(functions_content.contains("fn multiply"));
}

#[test]
fn test_trait_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("## test_crate::MyTrait"));
    assert!(lib_content.contains("**Type:** Trait"));
}

#[test]
fn test_methods_and_impls() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("**Methods:**"));
}

#[test]
fn test_constants_and_type_aliases() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("test_crate.md").expect("lib module not found");

    assert!(lib_content.contains("MAX_SIZE"));
    assert!(lib_content.contains("Result"));
}

#[test]
fn test_include_private() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");

    let output_public = converter::convert_to_markdown_multifile(&crate_data, false)
        .expect("Failed to convert to markdown");
    let output_private = converter::convert_to_markdown_multifile(&crate_data, true)
        .expect("Failed to convert to markdown");

    let public_count = output_public.files.len();
    let private_count = output_private.files.len();

    assert!(private_count >= public_count,
        "Private output should have at least as many files as public output");
}

#[test]
fn test_conversion_end_to_end() {
    let output_dir = std::env::temp_dir().join("cargo_doc_md_test");
    let _ = std::fs::remove_dir_all(&output_dir);

    let json_path = Path::new("tests/fixtures/test_crate.json");
    let options = ConversionOptions {
        input_path: json_path,
        output_dir: &output_dir,
        include_private: false,
    };

    cargo_doc_md::convert_json_file(&options).expect("Conversion failed");

    assert!(output_dir.join("test_crate").exists());
    assert!(output_dir.join("test_crate/index.md").exists());

    std::fs::remove_dir_all(&output_dir).ok();
}

#[test]
fn test_legacy_single_file_output() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let markdown = converter::convert_to_markdown(&crate_data, false)
        .expect("Failed to convert to markdown");

    assert!(markdown.contains("# test_crate"));
    assert!(markdown.contains("## Table of Contents"));
    insta::assert_snapshot!("legacy_single_file", markdown);
}
