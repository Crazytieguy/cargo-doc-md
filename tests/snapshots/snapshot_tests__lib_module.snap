---
source: tests/snapshot_tests.rs
expression: lib_content
---
**test_crate**

# Module: test_crate

## Contents

**Modules**

- [`async_example`](#async_example)
- [`errors`](#errors)
- [`functions`](#functions) - Function examples demonstrating various signatures and patterns.
- [`lifetimes`](#lifetimes)
- [`nested`](#nested) - Demonstrates nested module hierarchies.
- [`patterns`](#patterns)
- [`traits`](#traits)
- [`types`](#types) - Type definitions and containers.

**Structs**

- [`BoundedGeneric`](#boundedgeneric)
- [`Error`](#error)
- [`GenericStruct`](#genericstruct)
- [`PlainStruct`](#plainstruct)
- [`TupleStruct`](#tuplestruct)
- [`UnitStruct`](#unitstruct)

**Enums**

- [`ComplexEnum`](#complexenum)
- [`GenericEnum`](#genericenum)
- [`SimpleEnum`](#simpleenum)

**Functions**

- [`const_function`](#const_function)
- [`function_with_args`](#function_with_args)
- [`function_with_result`](#function_with_result)
- [`generic_function`](#generic_function)
- [`multiple_bounds`](#multiple_bounds)
- [`simple_function`](#simple_function)
- [`unsafe_function`](#unsafe_function) - An unsafe function that dereferences a raw pointer.

**Traits**

- [`DisplayDebug`](#displaydebug)
- [`MyTrait`](#mytrait)

**Constants**

- [`MAX_SIZE`](#max_size)
- [`MIN_SIZE`](#min_size)
- [`VERSION`](#version)

**Type Aliases**

- [`GenericResult`](#genericresult)
- [`Result`](#result)

---

## test_crate::BoundedGeneric

**Type:** Struct

**Generic Parameters:**
- T

**Fields:**

| Name | Type | Description |
|------|------|-------------|
| `data` | `T` | - |

**Methods:**

- `fn new(data: T) -> Self` - 
- `fn clone_data(self: &Self) -> T` - 



## test_crate::ComplexEnum

**Type:** Enum

**Variants:**

| Variant | Kind | Description |
|---------|------|-------------|
| `Unit` | Unit | - |
| `Tuple` | Tuple(String, i32) | - |
| `Struct` | Struct (2 fields) | - |

**Methods:**

- `fn name(self: &Self) -> Option<&str>` - 



## test_crate::DisplayDebug

**Type:** Trait

**Methods:**

- `format_both`



## test_crate::Error

**Type:** Struct

**Methods:**

- `fn new<impl Into<String>>(message: impl Trait) -> Self` - 
- `fn message(self: &Self) -> &str` - 

**Implemented Traits:** Error

**Trait Implementations:**

- **Display**
  - `fn fmt(self: &Self, f: & mut fmt::Formatter) -> fmt::Result` - 
- **Debug**
  - `fn fmt(self: &Self, f: & mut $crate::fmt::Formatter) -> $crate::fmt::Result` - 



## test_crate::GenericEnum

**Type:** Enum

**Generic Parameters:**
- T
- E

**Variants:**

| Variant | Kind | Description |
|---------|------|-------------|
| `Ok` | Tuple(T) | - |
| `Err` | Tuple(E) | - |
| `None` | Unit | - |

**Methods:**

- `fn is_ok(self: &Self) -> bool` - 
- `fn is_err(self: &Self) -> bool` - 
- `fn ok(self: Self) -> Option<T>` - 

**Trait Implementations:**

- **Debug**
  - `fn fmt(self: &Self, f: & mut $crate::fmt::Formatter) -> $crate::fmt::Result` - 
- **Clone**
  - `fn clone(self: &Self) -> GenericEnum<T, E>` - 



## test_crate::GenericResult

**Type:** Type Alias



## test_crate::GenericStruct

**Type:** Struct

**Generic Parameters:**
- T
- U

**Fields:**

| Name | Type | Description |
|------|------|-------------|
| `first` | `T` | - |
| `second` | `U` | - |

**Methods:**

- `fn new(first: T, second: U) -> Self` - 
- `fn swap(self: Self) -> GenericStruct<U, T>` - 
- `fn map_first<F, R>(self: Self, f: F) -> GenericStruct<R, U>` - 
- `fn duplicate(self: &Self) -> (T, U)` - 

**Trait Implementations:**

- **Debug**
  - `fn fmt(self: &Self, f: & mut $crate::fmt::Formatter) -> $crate::fmt::Result` - 
- **Clone**
  - `fn clone(self: &Self) -> GenericStruct<T, U>` - 



## test_crate::MAX_SIZE

**Type:** Constant



## test_crate::MIN_SIZE

**Type:** Constant



## test_crate::MyTrait

**Type:** Trait

**Methods:**

- `required_method`
- `provided_method`
- `another_provided`



## test_crate::PlainStruct

**Type:** Struct

**Fields:**

| Name | Type | Description |
|------|------|-------------|
| `name` | `String` | - |
| `value` | `i32` | - |

**Methods:**

- `fn new(name: String, value: i32) -> Self` - 
- `fn with_private(name: String, value: i32, private_field: bool) -> Self` - 
- `fn get_value(self: &Self) -> i32` - 
- `fn set_value(self: & mut Self, value: i32)` - 

**Implemented Traits:** Eq, StructuralPartialEq

**Trait Implementations:**

- **Clone**
  - `fn clone(self: &Self) -> PlainStruct` - 
- **PartialEq**
  - `fn eq(self: &Self, other: &PlainStruct) -> bool` - 
- **Default**
  - `fn default() -> Self` - 
- **Debug**
  - `fn fmt(self: &Self, f: & mut $crate::fmt::Formatter) -> $crate::fmt::Result` - 
- **MyTrait**
  - `fn required_method(self: &Self) -> String` - 
  - `fn provided_method(self: &Self) -> i32` - 



## test_crate::Result

**Type:** Type Alias



## test_crate::SimpleEnum

**Type:** Enum

**Variants:**

| Variant | Kind | Description |
|---------|------|-------------|
| `VariantA` | Unit | - |
| `VariantB` | Unit | - |
| `VariantC` | Unit | - |

**Methods:**

- `fn default_variant() -> Self` - 
- `fn is_variant_a(self: &Self) -> bool` - 

**Implemented Traits:** StructuralPartialEq, Eq

**Trait Implementations:**

- **PartialEq**
  - `fn eq(self: &Self, other: &SimpleEnum) -> bool` - 
- **Debug**
  - `fn fmt(self: &Self, f: & mut $crate::fmt::Formatter) -> $crate::fmt::Result` - 
- **Clone**
  - `fn clone(self: &Self) -> SimpleEnum` - 



## test_crate::TupleStruct

**Type:** Struct

**Tuple Struct** with 2 field(s)



## test_crate::UnitStruct

**Type:** Struct

**Unit Struct**



## test_crate::VERSION

**Type:** Constant



## Module: async_example



## test_crate::const_function

**Type:** Function

```rust
fn const_function(x: i32) -> i32
```



## Module: errors



## test_crate::function_with_args

**Type:** Function

```rust
fn function_with_args(name: &str, value: i32) -> String
```



## test_crate::function_with_result

**Type:** Function

```rust
fn function_with_result(value: i32) -> Result<String>
```



## Module: functions

Function examples demonstrating various signatures and patterns.

This module shows:
- Simple functions
- Generic functions with trait bounds
- Async functions
- Unsafe functions
- Const functions
- Higher-order functions



## test_crate::generic_function

**Type:** Function

```rust
fn generic_function<T>(item: T) -> String
```



## Module: lifetimes



## test_crate::multiple_bounds

**Type:** Function

```rust
fn multiple_bounds<T>(item: T) -> String
```



## Module: nested

Demonstrates nested module hierarchies.

This shows how documentation is structured for deeply nested modules.



## Module: patterns



## test_crate::simple_function

**Type:** Function

```rust
fn simple_function()
```



## Module: traits



## Module: types

Type definitions and containers.

This module demonstrates various type definitions including:
- Generic containers
- Enums with multiple variants
- Type aliases
- Structs with lifetime parameters



## test_crate::unsafe_function

**Type:** Function

An unsafe function that dereferences a raw pointer.

# Safety

The caller must ensure that `ptr` is valid, properly aligned,
and points to initialized memory.

```rust
fn unsafe_function(ptr: *const u8) -> u8
```
